pkg/scheduler/plugins/deviceshare/device_cache.go

```diff
@@ -0,0 +1,321 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	quotav1 "k8s.io/apiserver/pkg/quota/v1"
+	"k8s.io/klog/v2"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+	schedulingv1alpha1 "github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+)
+
+type nodeDeviceCache struct {
+	lock sync.Mutex
+	// nodeDeviceInfos stores nodeDevice for each node
+	// and uses node name as map key.
+	nodeDeviceInfos map[string]*nodeDevice
+}
+
+type nodeDevice struct {
+	lock        sync.Mutex
+	deviceTotal map[schedulingv1alpha1.DeviceType]deviceResources
+	deviceFree  map[schedulingv1alpha1.DeviceType]deviceResources
+	deviceUsed  map[schedulingv1alpha1.DeviceType]deviceResources
+	allocateSet map[schedulingv1alpha1.DeviceType]map[types.UID]*corev1.Pod
+}
+
+// deviceResources is used to present resources per device.
+// we use the minor of device as key
+// "0": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+// "1": {koordinator.sh/gpu-core:100, koordinator.sh/gpu-memory-ratio:100, koordinator.sh/gpu-memory: 16GB}
+type deviceResources map[int]corev1.ResourceList
+
+func newNodeDeviceCache() *nodeDeviceCache {
+	return &nodeDeviceCache{
+		nodeDeviceInfos: make(map[string]*nodeDevice),
+	}
+}
+
+func newNodeDevice() *nodeDevice {
+	return &nodeDevice{
+		deviceTotal: make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		deviceFree:  make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		deviceUsed:  make(map[schedulingv1alpha1.DeviceType]deviceResources),
+		allocateSet: make(map[schedulingv1alpha1.DeviceType]map[types.UID]*corev1.Pod),
+	}
+}
+
+func (in deviceResources) DeepCopy() deviceResources {
+	if in == nil {
+		return nil
+	}
+	out := deviceResources{}
+	for k, v := range in {
+		out[k] = v.DeepCopy()
+	}
+	return out
+}
+
+func (n *nodeDeviceCache) getNodeDevice(nodeName string) *nodeDevice {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	return n.nodeDeviceInfos[nodeName]
+}
+
+func (n *nodeDeviceCache) createNodeDevice(nodeName string) *nodeDevice {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.nodeDeviceInfos[nodeName] = newNodeDevice()
+	return n.nodeDeviceInfos[nodeName]
+}
+
+func (n *nodeDevice) resetDeviceTotal(resources map[schedulingv1alpha1.DeviceType]deviceResources) {
+	n.deviceTotal = resources
+	for deviceType := range resources {
+		n.resetDeviceFree(deviceType)
+	}
+}
+
+// updateCacheUsed is used to update deviceUsed when there is a new pod created/deleted
+func (n *nodeDevice) updateCacheUsed(deviceAllocations apiext.DeviceAllocations, pod *corev1.Pod, add bool) {
+	if len(deviceAllocations) > 0 {
+		for deviceType, allocations := range deviceAllocations {
+			if _, ok := n.allocateSet[deviceType][pod.UID]; ok && add {
+				// for non-failover scenario, pod might already exist in cache after Reserve step.
+				break
+			}
+			for _, allocation := range allocations {
+				n.updateDeviceUsed(deviceType, allocation, add)
```

Comment:
updateDeviceUsed/resetDeviceFree/updateAllocateSet should be unified for the variable level,  updateDeviceUsed is concrete to minor level, resetDeviceFree is concrete to device level, updateAllocateSet is concrete to whole deviceTypes.  I prefer deviceType by the exist for deviceType loop.